### ISSUE TITLE
Broken qhelp links

### ISSUE THREAD

**Issue by smowton** [github/codeql#4379]
I noticed a link to OWASP was broken in some Java qhelp, and figured I'd survey what links are broken. Here are the non-200 codes I got when trying to retrieve pages linked from qhelp (strikethrough indicates fixed):

| URL | Response code | Linked from |
|--|--|--|
| `http://7thgen.info/wiki/Code_Quality_For_Software_Architects` | 404 | `java/ql/src/Metrics/RefTypes/TEfferentSourceCoupling.qhelp ` |
| `http://docs.python-requests.org/en/master/user/advanced/#ssl-cert-verification` | 404 | `python/ql/src/Security/CWE-295/RequestWithoutValidation.qhelp ` |
| `http://infohost.nmt.edu/tcc/help/pubs/python/web/global-statement.html` | refused | `python/ql/src/Variables/ShadowGlobal.qhelp ` |
| `http://mongers.org/industrial-c++/` | timeout | `cpp/ql/src/Critical/ReturnValueIgnored.qhelp cpp/ql/src/jsf/4.24 Control Flow Structures/AV Rule 189.qhelp cpp/ql/src/jsf/4.24 Control Flow Structures/AV Rule 201.qhelp cpp/ql/src/Best Practices/UseOfGoto.qhelp cpp/ql/src/Best Practices/SloppyGlobal.qhelp cpp/ql/src/Best Practices/Magic Constants/MagicConstantsNumbers.qhelp cpp/ql/src/Best Practices/Magic Constants/MagicConstantsString.qhelp cpp/ql/src/Likely Bugs/Likely Typos/MissingEnumCaseInSwitch.qhelp ` |
| `http://selfcontained.us/2012/05/08/node-js-circular-dependencies/` | 403 | `javascript/ql/src/NodeJS/CyclicImport.qhelp ` |
| `http://www.javaworld.com/javaworld/jw-02-2006/jw-0227-control.html?page=1` | 404 | `java/ql/src/Likely Bugs/Serialization/IncorrectSerialVersionUID.qhelp ` |
| `http://www.leepoint.net/JavaBasics/methods/methods-22-local-variables.html` | nxdomain | `java/ql/src/Violations of Best Practice/legacy/ParameterAssignment.qhelp ` |
| `http://www2.sys-con.com/itsg/virtualcd/java/archives/0605/ford/index.html` | 404 | `java/ql/src/Likely Bugs/Frameworks/Swing/ThreadSafety.qhelp ` |
| `http://www4.in.tum.de/~juergens/publications/ICSE2009_RP_0110_juergens.pdf` | 404 | `python/ql/src/external/MostlyDuplicateFile.qhelp python/ql/src/external/MostlySimilarFile.qhelp python/ql/src/external/MostlyDuplicateClass.qhelp python/ql/src/external/SimilarFunction.qhelp ` |
| `https://blog.liftsecurity.io/2015/01/14/the-dangers-of-square-bracket-notation#lift-security` | 403 | `javascript/ql/src/Security/CWE-400/RemotePropertyInjection.qhelp ` |
| `https://docs.microsoft.com/en-us/windows/desktop/Intl/era-handling-for-the-japanese-calendar/` | 404 | `csharp/ql/src/Likely Bugs/MishandlingJapaneseEra.qhelp ` |
| `https://github.com/microsoft/TypeScript/blob/master/doc/spec.md#3.8.9` | 404 | `javascript/ql/src/Declarations/SuspiciousMethodNameDeclaration.qhelp ` |
| `https://github.com/microsoft/TypeScript/blob/master/doc/spec.md#8.3.1` | 404 | `javascript/ql/src/Declarations/SuspiciousMethodNameDeclaration.qhelp ` |
| `https://lodash.com/docs/latest#unescape` | 404 | `javascript/ql/src/Security/CWE-611/Xxe.qhelp ` |
| `https://owasp.org/www-project-web-security-testing-guide/latest/4-Web_Application_Security_Testing/07-Input_Validation_Testing/18-Testing_for_Server_Side_Template_Injection` | 404 | `javascript/ql/src/experimental/Security/CWE-94/ServerSideTemplateInjection.qhelp ` |
| `https://wiki.sei.cmu.edu/confluence/display/c/EXP34-C.+Do+not+dereference+null+pointerss` | 404 | `cpp/ql/src/Critical/MissingNullTest.qhelp ` |
| `https://www.npmjs.org/doc/files/package.json.html` | 404 | `javascript/ql/src/NodeJS/UnusedDependency.qhelp ` |
| ~`https://www.owasp.org/index.php/Information_Leak_(information_disclosure)`~ | 404 | `python/ql/src/Security/CWE-209/StackTraceExposure.qhelp java/ql/src/Security/CWE/CWE-209/StackTraceExposure.qhelp csharp/ql/src/Security Features/CWE-209/ExceptionInformationExposure.qhelp javascript/ql/src/Security/CWE-209/StackTraceExposure.qhelp ` |
| ~`https://www.owasp.org/index.php/Path_traversal`~ | 404 | `python/ql/src/Security/CWE-022/PathInjection.qhelp python/ql/src/Security/CWE-022/TarSlip.qhelp java/ql/src/Security/CWE/CWE-022/TaintedPath.qhelp java/ql/src/Security/CWE/CWE-022/ZipSlip.qhelp cpp/ql/src/Security/CWE/CWE-022/TaintedPath.qhelp csharp/ql/src/experimental/CWE-099/TaintedWebClient.qhelp csharp/ql/src/Security Features/CWE-022/TaintedPath.qhelp csharp/ql/src/Security Features/CWE-022/ZipSlip.qhelp javascript/ql/src/Security/CWE-022/TaintedPath.qhelp javascript/ql/src/Security/CWE-022/ZipSlip.qhelp ` |
| ~`https://www.owasp.org/index.php?title=Testing_for_XPath_Injection_(OTG-INPVAL-010)`~ | 404 | `java/ql/src/experimental/Security/CWE/CWE-643/XPathInjection.qhelp csharp/ql/src/Security Features/CWE-643/XPathInjection.qhelp javascript/ql/src/Security/CWE-643/XpathInjection.qhelp ` |
| `https://www.securecoding.cert.org/confluence/display/cplusplus/DCL07-CPP.+Minimize+the+scope+of+variables+and+methods` | 404 | `cpp/ql/src/Best Practices/Unused Entities/UnusedStaticVariables.qhelp ` |
| `http://blogs.msdn.com/b/ericlippert/archive/2009/11/12/closing-over-the-loop-variable-considered-harmful.aspx` | 404 | `csharp/ql/src/Language Abuse/ForeachCapture.qhelp ` |
| `http://blogs.msdn.com/b/vijaysk/archive/2008/03/19/object-referenceequals-valuevar-valuevar-will-always-return-false.aspx` | 404 | `csharp/ql/src/Likely Bugs/ReferenceEqualsOnValueTypes.qhelp ` |
| `http://blogs.msdn.com/oldnewthing/archive/2004/05/07/127826.aspx` | 404 | `cpp/ql/src/Likely Bugs/OO/NonVirtualDestructor.qhelp ` |
| `https://blogs.msdn.microsoft.com/prashant_upadhyay/2011/07/14/why-debugfalse-in-asp-net-applications-in-production-environment/` | 404 | `csharp/ql/src/Security Features/CWE-011/ASPNetDebug.qhelp ` |

In addition, we link at least these MSDN pages most of which (unsure why most but not all) now produce a pretty unfriendly "retired documentation" page that invites the user to download a PDF of the old docs, and should probably be bumped forwards to a more recent Visual Studio version where the documentation continues to be relevant:

```
http://msdn.microsoft.com/en-US/library/k68ktdwf%28v=VS.80%29.aspx
http://msdn.microsoft.com/en-gb/library/b34dt9cd%28v=vs.80%29.aspx
http://msdn.microsoft.com/en-us/library/126fe14k%28v=VS.80%29.aspx
http://msdn.microsoft.com/en-us/library/36k2cdd4%28v=VS.80%29.aspx
http://msdn.microsoft.com/en-us/library/503x3e3s%28v=vs.80%29.aspx
http://msdn.microsoft.com/en-us/library/8y82wx12%28v=VS.80%29.aspx
http://msdn.microsoft.com/en-us/library/bsc2ak47(v=vs.80).aspx
http://msdn.microsoft.com/en-us/library/e59b22c5(v=vs.80).aspx
http://msdn.microsoft.com/en-us/library/k0t5wee3%28v=VS.80%29.aspx
http://msdn.microsoft.com/en-us/library/ms173147%28VS.80%29.aspx
http://msdn.microsoft.com/en-us/library/ms173147(v=vs.80).aspx
http://msdn.microsoft.com/en-us/library/wa80x488%28VS.80%29.aspx
http://msdn.microsoft.com/en-us/library/wcz57btd%28v=vs.80%29.aspx
http://msdn.microsoft.com/en-us/library/x9fsa0sw(v=vs.80).aspx
http://msdn.microsoft.com/en-us/library/yh598w02(v=vs.80).aspx
https://msdn.microsoft.com/en-gb/library/b34dt9cd%28v=vs.80%29.aspx
https://msdn.microsoft.com/en-us/library/89211k9b(VS.80).aspx
https://msdn.microsoft.com/en-us/library/ms254947(v=vs.80).aspx
http://msdn.microsoft.com/en-us/library/03b5270t(v=vs.71).aspx
http://msdn.microsoft.com/en-us/library/2dzy4k6e%28v=vs.71%29.aspx
http://msdn.microsoft.com/en-us/library/4s7x1k91%28v=VS.71%29.aspx
http://msdn.microsoft.com/en-us/library/98f28cdx(v=vs.71).aspx
http://msdn.microsoft.com/en-us/library/aa246471%28v=VS.60%29.aspx
http://msdn.microsoft.com/en-us/library/aa645739(v=vs.71).aspx
http://msdn.microsoft.com/en-us/library/aa691135(v=vs.71).aspx
http://msdn.microsoft.com/en-us/library/ah19swz4(v=vs.71).aspx
http://msdn.microsoft.com/en-us/library/cscsdfbt(v=vs.71).aspx
http://msdn.microsoft.com/en-us/library/kxszd0kx(v=vs.71).aspx
http://msdn.microsoft.com/en-us/library/sbf85k1c(v=vs.71).aspx
http://msdn.microsoft.com/en-us/library/scekt9xw(v=vs.71).aspx
http://msdn.microsoft.com/en-us/library/ty2ax9z9%28v=vs.71%29.aspx
http://msdn.microsoft.com/en-us/library/wc7014hz%28VS.71%29.aspx
```

**Comment by geoffw0** [comment]
It looks like this has been fixed, can you confirm?

**Comment by smowton** [comment]
As of today, the list of broken links appears to be:

```
FAIL http://blogs.msdn.com/b/ericlippert/archive/2009/11/12/closing-over-the-loop-variable-considered-harmful.aspx
FAIL http://blogs.msdn.com/b/vijaysk/archive/2008/03/19/object-referenceequals-valuevar-valuevar-will-always-return-false.aspx
FAIL https://blogs.msdn.microsoft.com/prashant_upadhyay/2011/07/14/why-debugfalse-in-asp-net-applications-in-production-environment/
FAIL http://7thgen.info/wiki/Code_Quality_For_Software_Architects
FAIL http://effbot.org/zone/default-values.htm
FAIL http://effbot.org/zone/import-confusion.htm
FAIL http://effbot.org/zone/python-with-statement.htm
FAIL http://infohost.nmt.edu/tcc/help/pubs/python/web/global-statement.html
FAIL http://mongers.org/industrial-c++/
FAIL http://selfcontained.us/2012/05/08/node-js-circular-dependencies/
FAIL http://troll.cs.ua.edu/ACP-PY/index_13.html
FAIL http://wiki.netbeans.org/Java_Hints
FAIL http://www.ecma-international.org/ecma-262/6.0
FAIL http://www.ibm.com/developerworks/java/library/j-jtp0618/index.html#4
FAIL http://www.ibm.com/developerworks/library/j-eaed6/
FAIL http://www.leepoint.net/JavaBasics/methods/methods-22-local-variables.html
FAIL http://www2.sys-con.com/itsg/virtualcd/java/archives/0605/ford/index.html
FAIL http://www4.in.tum.de/~juergens/publications/ICSE2009_RP_0110_juergens.pdf
FAIL https://blog.liftsecurity.io/2015/01/14/the-dangers-of-square-bracket-notation#lift-security
FAIL https://docs.microsoft.com/en-us/windows/desktop/Intl/era-handling-for-the-japanese-calendar/
FAIL https://github.com/microsoft/TypeScript/blob/master/doc/spec.md#3.8.9
FAIL https://github.com/microsoft/TypeScript/blob/master/doc/spec.md#8.3.1
FAIL https://lodash.com/docs/latest#unescape
FAIL https://media.blackhat.com/eu-13/briefings/Osipov/bh-eu-13-XML-data-osipov-slides.pdf
FAIL https://nodesecurity.io/advisories/566
FAIL https://owasp.org/www-community/attacks/Man-in-the-middle_attack
FAIL https://wiki.hsr.ch/Prog3/files/overload72-FINAL_DesigningHeaderFiles.pdf
FAIL https://www.ibm.com/developerworks/java/library/j-jtp01274/index.html#3.2
FAIL https://www.ibm.com/developerworks/java/library/j-jtp03216/index.html
FAIL https://www.ibm.com/developerworks/java/library/j-jtp05273/index.html
FAIL https://www.npmjs.org/doc/files/package.json.html
FAIL https://www.owasp.org/index.php/Information_Leak_(information_disclosure)
FAIL https://www.owasp.org/index.php/XSS_%28Cross_Site_Scripting%29_Prevention_Cheat_Sheet
FAIL https://www.rsaconference.com/writable/presentations/file_upload/asd-f03-serial-killer-silently-pwning-your-java-endpoints.pdf
FAIL https://www.vsecurity.com//download/publications/XMLDTDEntityAttacks.pdf
FAIL https://xunit.github.io/
```

**Comment by intrigus-lgtm** [comment]
http://www4.in.tum.de/~juergens/publications/ICSE2009_RP_0110_juergens.pdf -> https://wwwbroy.in.tum.de/~juergens/publications/ICSE2009_RP_0110_juergens.pdf
https://nodesecurity.io/advisories/566 -> https://npmjs.com/advisories/566

**Comment by smowton** [comment]
Ah screw it, it's Friday: https://github.com/github/codeql/pull/5757

(Thanks @intrigus-lgtm I've included those two)

**Comment by smowton** [comment]
Fixed in #5757
